src/StreamDeckLib/ConnectionManager.cs

```diff
@@ -1,300 +1,360 @@
-		private BaseStreamDeckPlugin _Plugin;
-		private static ILoggerFactory _LoggerFactory;
-		private static ILogger _Logger;
-
-		void Dispose(bool disposing)
-		{
-			if (!disposedValue)
-			{
-				if (disposing)
-				{
-					_Proxy.Dispose();
-				}
-
-				disposedValue = true;
-			}
-		}
-
-		// This code added to correctly implement the disposable pattern.
-		public void Dispose()
-		{
-			// Do not change this code. Put cleanup code in Dispose(bool disposing) above.
-			Dispose(true);
-			// TODO: uncomment the following line if the finalizer is overridden above.
-			// GC.SuppressFinalize(this);
-		}
-		#endregion
-	}
-}
+﻿using McMaster.Extensions.CommandLineUtils;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Logging.Abstractions;
+using Newtonsoft.Json;
+using StreamDeckLib.Messages;
+using System;
+using System.ComponentModel.DataAnnotations;
+using System.Diagnostics;
+using System.IO;
+using System.Net.WebSockets;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace StreamDeckLib
+{
+    /// <summary>
+    /// This class manages the connection to the StreamDeck hardware
+    /// </summary>
+    public partial class ConnectionManager : IDisposable
+    {
+        private int _Port;
+        private string _Uuid;
+        private string _RegisterEvent;
+		private IStreamDeckProxy _Proxy;
+        private ConnectionManager() { }
+        public Messages.Info Info { get; private set; }
+        public static ConnectionManager Initialize(string[] commandLineArgs,
+				ILoggerFactory loggerFactory = null, IStreamDeckProxy streamDeckProxy = null)
+        {
+			var waitForDebugger = false;
+            using (var app = new CommandLineApplication())
+            {
+                app.HelpOption();
+
+                var optionPort = app.Option<int>("-port|--port <PORT>", "The port the Elgato StreamDeck software is listening on", CommandOptionType.SingleValue);
+                var optionPluginUUID = app.Option("-pluginUUID <UUID>", "The UUID that the Elgato StreamDeck software knows this plugin as.", CommandOptionType.SingleValue);
+                var optionRegisterEvent = app.Option("-registerEvent <REGEVENT>", "The registration event", CommandOptionType.SingleValue);
+                var optionInfo = app.Option("-info <INFO>", "Some information", CommandOptionType.SingleValue);
+
+#if DEBUG
+				var optionDebug = app.Option<bool>("-d|--debug", "Wait for a debugger to attach before execution", CommandOptionType.NoValue);
+#endif
+
+                app.Parse(commandLineArgs);
+
+				// Validate that the command line parameters are valid, and display the help if they aren't.
+				if (app.GetValidationResult() != ValidationResult.Success)
+				{
+					app.ShowHelp();
+					throw new ArgumentException("The command line parameters were not valid or could not be parsed.");
+				}
+				
+#if DEBUG
+				waitForDebugger = optionDebug.HasValue() && optionDebug.ParsedValue;

```

Comment:
so is the idea that it spins up, then we'll manually connect the debugger at some point?

Reply:
Correct. The variable is passed into the plugin's `Initialize()` method, and within that method, the first order of business is to go in a wait loop until either a) a debugger is attached, or b) the cancellation token's source has requested cancellation - so that the thread doesn't hang around indefinitely.

This is all done *before* any websocket connection attempts to give the developer an opportunity to monitor every step of the communication process between the plugin and its caller, aside from the command line invocation.